src/segmantic/seg/monai_unet.py

```diff
@@ -548,3 +700,100 @@ def print_table(header, vals, indent="\t"):
                 confusion_metrics,
                 (np.squeeze(x.cpu().numpy()) for x in conf_matrix.aggregate()),
             )
+
+
+def cross_validate(
+    image_dir: Path,
+    labels_dir: Path,
+    tissue_list: Path,
+    output_dir: Path,
+    config_files_dir: Path,
+    checkpoint_file: Path = None,
+    max_epochs: int = 100,
+    generalize_test: bool = False,
+    generalize_img_dir: Optional[Path] = None,
+    generalize_lbl_dir: Optional[Path] = None,
+    n_splits: int = 7,
+    save_nifti: bool = True,
+    gpu_ids: List[int] = [0],
+):
+    print_config()
+    print("Cross-validating")
+    output_dir = Path(output_dir)
+    output_dir.mkdir(exist_ok=True)
+
+    tissue_dict = load_tissue_list(tissue_list)
+    print(tissue_dict)
+
+    data_dicts = PairedDataSet.create_data_dict(
+        image_dir=image_dir, labels_dir=labels_dir
+    )
+    all_datafold_paths: List[Path] = PairedDataSet.kfold_crossval(
+        num_splits=n_splits, data_dicts=data_dicts, output_dir=output_dir
+    )
+
+    for config_file in Path(config_files_dir).iterdir():
+        # ToDo: add yaml file support
+        assert config_file.suffix in [".json"]
+
+        output_dir_scenario = output_dir.joinpath(str(config_file.name))
+        if not output_dir_scenario.exists():
+            output_dir_scenario.mkdir()
+
+        for count, dataset_path in enumerate(all_datafold_paths):
+
+            current_output = output_dir_scenario.joinpath(str(count))
+            print(current_output)
+
+            current_output.mkdir(exist_ok=True)
+
+            data = json.loads(config_file.read_text())
+            data["dataset"] = str(dataset_path)
+            data["output_dir"] = str(current_output)
+            current_layers = data["layers"]
+            current_strides = data["strides"]
+            with open(config_file, "w") as f:
+                json.dump(data, f, indent=4)
+
+            print("start training")
+
+            result = sp.run(
+                [
+                    sys.executable,
+                    str(Path.cwd().parents[2].joinpath("scripts/run_monai_unet.py")),
+                    "train-config",
+                    "-c",
+                    config_file,
+                ]
+            )
+            print(result)
+            print("training finished")
+            if (
+                generalize_test
+                and generalize_img_dir is not None
+                and generalize_lbl_dir is not None
+            ):
+                assert generalize_img_dir.is_dir() and generalize_lbl_dir.is_dir()
+
+                test_images = sorted(list(generalize_img_dir.glob(".nii.gz")))
+                test_labels = sorted(list(generalize_lbl_dir.glob(".nii.gz")))
+
+                assert len(test_images) == len(test_labels)
+
+                for file in current_output.iterdir():
+                    if file.match("*.ckpt"):
```

Comment:
not sure you answered if there are multiple `.ckpt` files in the `current_output`.

Reply:
interesting. is there any trend or can the second best easily have a better test dice than the *best*?